pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1260,26 +1263,48 @@ func NewArtifactSearchQuery() *ArtifactSearchQuery {
 	return &q
 }
 
+func agcsQueryHelper(q *ArtifactSearchQuery) bool {
+	for _, val := range q.ArtifactGCStrategies {
+		if val == true {
+			return val
+		}
+	}
+	return false
+}
+
 func (w *Workflow) SearchArtifacts(q *ArtifactSearchQuery) Artifacts {
 
 	var artifacts Artifacts
 
 	for _, n := range w.Status.Nodes {
 		t := w.GetTemplateByName(n.TemplateName)
 		for _, a := range n.GetOutputs().GetArtifacts() {
-			artifactByName := t.GetOutputs().GetArtifactByName(a.Name)
-			templateStrategy := t.GetArtifactGC().GetStrategy()
-			wfStrategy := w.Spec.GetArtifactGC().GetStrategy()
-			strategy := wfStrategy
-			if templateStrategy != ArtifactGCNever {
-				strategy = templateStrategy
+			toQuery := true
+			if agcsQueryHelper(q) == true {
+				templateStrategy := t.GetArtifactGC().GetStrategy()
+				wfStrategy := w.Spec.GetArtifactGC().GetStrategy()
+				strategy := wfStrategy
+				if templateStrategy != ArtifactGCNever {
+					strategy = templateStrategy
+				}
+				if q.ArtifactGCStrategies[strategy] != true {
```

Comment:
minor - simplify ` q.ArtifactGCStrategies[strategy] != true` to ` !q.ArtifactGCStrategies[strategy]`